### PR TITLE
feat(grouping): Store primary basis for hash in `GroupHashMetadata` table

### PR DESCRIPTION
**Pull request by lobsterkatie**
This adds a new field, `hash_basis`, to the grouphash metadata table, to track the primary grouping method used when generating the hash. It also adds a new helper, `_get_hash_basis`, which gets the appropriate values based on the main contributing variant (a data structure containing grouping data used during the hashing process). The assigned values are as one would expect, with the caveat that hybrid fingerprints are categorized according to the original grouping method. (In other words, an event with a hybrid fingerprint which was originally grouped on stacktrace will be categorized as having been grouped by stacktrace. The hybrid-fingerprint-ness will be accounted for in the hashing metadata soon to be added to the table.) 

The tests here are adapted from the variant snapshot tests, and use the same inputs. The primary difference between the variant snapshots and these snapshots is that in these snapshots, only contributing variants are included, and within those variants, only contributing values are included. There is also an extra test, `test_unknown_hash_basis`, which fakes variant data rather than getting it from a real input, in order to test the handling of unexpected variant results.

### CODE REVIEW

File: src/sentry/grouping/ingest/grouphash_metadata.py [hunk 1]
```diff
@@ -1,24 +1,106 @@
...
+    "in-app thread stack-trace": HashBasis.STACKTRACE,
+    # The value in `message` or `log_entry`, such as from using `capture_message` or calling
+    # `capture_exception` on a string
+    "message": HashBasis.MESSAGE,
+    # Error type and value, in cases where all frames are ignored by stacktrace rules
+    "exception": HashBasis.MESSAGE,
+    # Error type and value, in cases where there's no stacktrace
+    "in-app exception": HashBasis.MESSAGE,
+    # Fingerprint set by the user, either in the client or using server-side fingerprinting rules
+    "custom fingerprint": HashBasis.FINGERPRINT,
```
Comment (JoshFerge) [inline src/sentry/grouping/ingest/grouphash_metadata.py:32]: is it possible at all to delineate between these two? (user-set client vs server fingerprint)

Comment (lobsterkatie) [reply src/sentry/grouping/ingest/grouphash_metadata.py:32]: That will be included in the `hashing_metadata` - there's a `fingerprint_source` entry.

**Review by wedamija — COMMENTED** [review]
Migration looks safe to me

File: src/sentry/grouping/ingest/grouphash_metadata.py [hunk 2]
```diff
@@ -1,24 +1,106 @@
...
+from sentry.eventstore.models import Event
+from sentry.grouping.variants import BaseVariant
 from sentry.models.grouphash import GroupHash
-from sentry.models.grouphashmetadata import GroupHashMetadata
+from sentry.models.grouphashmetadata import GroupHashMetadata, HashBasis
+from sentry.models.project import Project
+
+logger = logging.getLogger(__name__)
+
+GROUPING_METHODS_BY_DESCRIPTION = {
```
Comment (JoshFerge) [inline src/sentry/grouping/ingest/grouphash_metadata.py:13]: is there a place where all of these keys are defined?

**Review by JoshFerge — APPROVED** [review]
looks good. eyeballed a dozen snapshots and output looks great!

File: src/sentry/grouping/ingest/grouphash_metadata.py (hunk 2 above)
Comment (lobsterkatie) [reply src/sentry/grouping/ingest/grouphash_metadata.py:13]: No, because `description` is a property, and it behaves differently depending on the grouping method.

Sometimes it returns a hard-coded string, as the `CustomFingerprintVariant` does here:
https://github.com/getsentry/sentry/blob/81a16b22dea0f22c2f0c33c1887afedaf97596c6/src/sentry/grouping/variants.py#L166-L168

Other times, though, it returns a computed value, as with component variants (which are what's used when grouping by stacktrace, by message, by template, and by security report data), as can be seen here (the first is `ComponentVariant.description` and the second is `GroupingComponent.description`):
https://github.com/getsentry/sentry/blob/81a16b22dea0f22c2f0c33c1887afedaf97596c6/src/sentry/grouping/variants.py#L117-L119
https://github.com/getsentry/sentry/blob/81a16b22dea0f22c2f0c33c1887afedaf97596c6/src/sentry/grouping/component.py#L64-L83

Then there are spots where we modify the description, as we do in `SaltedComponentVariant` (which we use when we have a hybrid fingerprint) :
https://github.com/getsentry/sentry/blob/81a16b22dea0f22c2f0c33c1887afedaf97596c6/src/sentry/grouping/variants.py#L197-L199

So, in short... it's complicated.

Comment (JoshFerge) [reply src/sentry/grouping/ingest/grouphash_metadata.py:13]: got it. thank you for the explanation!
